Update all npm

jaredrethman opened the pull request:
### Description of the Change

* Updated all NPM dependencies.
* CleanWebpackPlugin named export change. Requiring destructing at assignment in latest update.
* Added `module` and `process` to `.eslint` globals.
* Replaced `.babel` runtime config with `bable.config.js`.
* These updates will require a minimum of Node v8.11:
   * This has been added to the README.md
   * Added `.npmrc` to enforce `engine-strict`. `engineStrict` via package.json property is deprecated and un-enforceable. 
* Added `local` property to `testing.urls` in package.json, resolving an error when running `npm run test:a11y` locally.

### Alternate Designs

N/A

### Benefits

- House keeping. 
- Lines the code base up for usage of most stable packages in future enhancements. For example:
   - Webpack Dev Server & Hot Module Replacement.
   - Jest & Enzyme unit testing.
- Fixed 2 x bugs.
- Replaced `.babelrc` with `babel.config.js`: 
   - Will compile `node_modules/` - https://babeljs.io/docs/en/configuration#whats-your-use-case.
   - ["The main benefit being that users can include JS logic to build up their config structures..."](https://babeljs.io/docs/en/config-files#javascript)
   - ["JS configs are great because they can compute a config on the fly"](https://babeljs.io/docs/en/config-files#apicache)
   - Up to date with [Babel Usage Guide](https://babeljs.io/docs/en/usage).
   - Support more complex configurations.
- Force `engineStrict|engine-strict` Node version requirements.
- `npm run test:a11y` can now be run locally.

### Possible Drawbacks

- NPM updates are on predominantly minor and patch versions. For package updates within major version updates, there are no breaking changes that would effect this code base (see below for list of these packages, and links, where possible, to release notes outlining breaking changes).

### Verification Process

- Have tested all `npm run *` commands from package.json
- Have tested transpilation of ES6 to ES5; [Array.prototype.reduce()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce), [Array.prototype.reduceRight()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/ReduceRight), [Array.prototype.includes()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes),  [String.prototype.includes()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes), [Array.prototype.map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map), [Template literals (Template strings)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals), [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) and [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax). This working as expected, and can be tested with [this](https://gist.github.com/jaredrethman/65268af9654b3e43e1d01a068bcc9347) snippet, which will produce the below output in the console, and transpile to [this](https://tinyurl.com/y5dtn8u9). 

![Screenshot 2019-10-14 at 20 38 28](https://user-images.githubusercontent.com/3676975/66774735-9b6a1e80-eec2-11e9-9fd0-68c6acfdf385.png)

- I was able to verify through running all commands, and ensured the command line stayed free from any warnings/errors.

- `.npmrc` to enforce `engineStrict`.  `engineStrict` property via `package.json` is [deprecated, since NPM 3](https://docs.npmjs.com/files/package.json#enginestrict).

<img width="901" alt="Screenshot 2019-10-27 at 14 06 11" src="https://user-images.githubusercontent.com/3676975/67634323-05bb8f80-f8c3-11e9-9921-8f2fdacfb4dc.png">

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

### Applicable Issues

**Style Lint:**

Not an issue to this update, but applicable to why stylelint package, and associated packages, haven't been updated.  

<img width="1222" alt="Screenshot 2019-10-27 at 03 00 53" src="https://user-images.githubusercontent.com/3676975/67627968-7bddd900-f866-11e9-993c-7d489f7f812b.png">

<img width="989" alt="Screenshot 2019-10-27 at 01 37 33" src="https://user-images.githubusercontent.com/3676975/67627342-c9544900-f85a-11e9-9bd7-e0c09012f828.png">

* Below is the current list of outdated packages. The current `stylelint*` config, is the only configuration, which produces no warnings/errors.

<img width="484" alt="Screenshot 2019-10-27 at 01 32 23" src="https://user-images.githubusercontent.com/3676975/67627314-010ec100-f85a-11e9-825c-859b6cffa62b.png">

* The below warning has been resolved.

<img width="1009" alt="Screenshot 2019-10-10 at 15 24 29" src="https://user-images.githubusercontent.com/3676975/66572909-1c977d80-eb72-11e9-9d7d-eece54475fcf.png">

**Major version changes:**
* eslint: v5 to v6
https://github.com/eslint/eslint/releases?after=v6.0.0
* eslint-loader: v2 to v3
https://github.com/webpack-contrib/eslint-loader/releases/tag/v3.0.0 
* babel-eslint: v9 to v10
https://github.com/babel/babel-eslint/releases/tag/v10.0.0
* clean-webpack-plugin: v2 to v3
https://github.com/johnagan/clean-webpack-plugin/releases/tag/v3.0.0
* lint-staged: v8 to v9
https://github.com/okonet/lint-staged/releases?after=v9.2.0
* postcss-preset-env: v5 to v6
https://github.com/csstools/postcss-preset-env/releases
* stylelint-config-wordpress: v13 to v14
https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress/releases
* stylelint-order: v1 to v2
https://github.com/hudochenkov/stylelint-order/releases
* terser: v3 to v4
https://github.com/terser/terser/blob/master/CHANGELOG.md#v400
* webpackbar: v3 to v4
https://github.com/nuxt/webpackbar/blob/master/CHANGELOG.md#400-2019-08-05

### Changelog Entry

- Updated all npm packages.
- Added `local` property to `testing.urls` in package.json, resolving an error when running `npm run test:a11y` locally.
- Replaced `.babelrc` with `babel.config.js`
- Updated CleanWebpackPlugin variable assignment, requiring destructing in the newer version.
- Added `.npmrc` to enforce `engineStrict`

Associated Plugin Scaffold PR - https://github.com/10up/plugin-scaffold/pull/102

jaredrethman committed:
Updated all NPM packages, regenerated package-lock.json

jaredrethman committed:
Replaced babelrc with babel.config.js version.

jaredrethman committed:
Added "module" to eslint globals property.

jaredrethman committed:
Destructing CleanWebpackPlugin assignment.

jaredrethman committed:
Node version flag for README

jaredrethman committed:
Regenerated package-lock. Added additional property to url property in package.json, to fix issue when running npm run test:a11y. Updated node engine version.

jaredrethman committed:
Merge remote-tracking branch 'upstream/master'

jaredrethman committed:
Merged upstream/master in. Resolved conflicts.

jaredrethman committed:
Updated stylelint-* packages

jaredrethman committed:
Added npmrc to force engine-strict.

jaredrethman committed:
Merge remote-tracking branch 'upstream/master'

jaredrethman committed:
Merge remote-tracking branch 'origin/master' into update-all-npm

# Conflicts:
#	package-lock.json

jaredrethman committed:
Node v8.11 minimum update

timwright12 commented:
@jaredrethman can you add these to the plugin repo as well? Also, what's the benefit you see of moving `babelrc`?

jaredrethman commented:
@timwright12 
> can you add these to the plugin repo as well?

Definitely, I'll get on that asap.

> what's the benefit you see of moving `.babelrc`

1.) Primary motivation: Babel >= 7 recommends `babel.config.js` over `.babelrc`. Future proofing.
2.) We can perform logical operations before deciding which presets/plugins to use i.e. `NODE_ENV` - https://github.com/10up/theme-scaffold/pull/149

More here - https://babeljs.io/docs/en/config-files

jaredrethman committed:
Minor v-bumps

jaredrethman commented:
@timwright12 have added this Code Review equivalent to the Plugin Scaffold - https://github.com/10up/plugin-scaffold/pull/102 and updated this pull.

timwright12 commented on package.json:
```diff
@@ -51,63 +51,63 @@
   "license": "MIT",
```
@jaredrethman something in here is breaking our CSS processing, can you investigate? You can test by adding:

```
body {
	background: var(--c-black);
	color: var(--c-white);

	@media (--bp-small) {
		background: var(--c-white);
		color: var(--c-black);
	}

}
```

into `/assets/css/base/index.css` and noting that the media queries don't work after npm install on this branch, but they work on master

samikeijonen replied:
@timwright12 & @jaredrethman. It's because `@custom-media` rules are inside `:root` which they shouldn't.

I think they were working before with luck:)

So to fix this, `media-queries.css` should look like this without `:root`.

```css
/*
 * Media Queries
 */
@custom-media --bp-tiny ( min-width: 25em ); /* 400px */
@custom-media --bp-small ( min-width: 30em ); /* 480px */
@custom-media --bp-medium ( min-width: 48em ); /* 768px */
@custom-media --bp-large ( min-width: 64em ); /* 1024px */
@custom-media --bp-xlarge ( min-width: 80em ); /* 1280px */
@custom-media --bp-xxlarge ( min-width: 90em ); /* 1440px */

/* WP Core Breakpoints (used for the admin bar for example) */
@custom-media --wp-small ( min-width: 600px );
@custom-media --wp-medium-max (max-width: 782px);
```

Jared, let me know if you have questions or need help.

jaredrethman replied:
@samikeijonen @timwright12 haven't had much time lately to fix/address this. The small investigations I did recently pointed to `post-css-env` as the culprit. Quick fix was to downgrade it back, but if you're suggestion works - that would be 1000% better.

Can look into this over the coming weekend, unless you're able to tackle it sooner :)

jaredrethman replied:
@samikeijonen your suggestion worked perfectly. @timwright12 let me know if you have success with the latest commit - https://github.com/10up/theme-scaffold/pull/162/commits/87f2abe906aad56c614c7debb4501b8658743ed9.

jaredrethman committed:
A few minor updates & un-scoped @custom-media from :root

timwright12 commented:
@jaredrethman looks like some conflicts popped up.

We'll need these updates for the plugin as well https://github.com/10up/plugin-scaffold/pull/102

One last note for `Added module and process to .eslint globals.` -- I don't see those in the updates in the `.eslintrc` file

tylercherpak commented:
@timwright12 given the nature of these changes, I think we should try to get them merged posthaste since it is very likely that these types of merge conflicts will keep coming up.

jaredrethman committed:
Merge remote-tracking branch 'upstream/master'

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

jaredrethman committed:
Merge remote-tracking branch 'origin/master' into update-all-npm

# Conflicts:
#	.eslintrc
#	package-lock.json
#	package.json

jaredrethman committed:
Adding module + process to eslint global whitelist

jaredrethman commented:
Hey @timwright12 ,

I have resolved those conflicts, updated https://github.com/10up/plugin-scaffold/pull/102 and added `module` & `process` to eslint global whitelist.

timwright12 commented:
@jaredrethman looks good, just confirming the plugin repo before merge
